psi4/src/psi4/libmints/matrix.cc

```diff
@@ -1975,6 +1967,71 @@ void Matrix::cholesky_factorize() {
             }
         }
     }
+    // Zero the upper triangle which hasn't been touched by LAPACK
+    zero_upper();
+}
+
+void Matrix::pivoted_cholesky(double tol, std::vector<std::vector<int>> &pivot) {
+    if (symmetry_) {
+        throw PSIEXCEPTION("Matrix::pivoted_cholesky: Matrix is non-totally symmetric.");
+    }
+
+    // Cholesky basis size
+    Dimension nchol(rowspi_);
+
+    // Pivot indices
+    pivot.clear();
+    pivot.resize(nirrep_);
+    // Work array for LAPACK
+    std::vector<double> work(2 * max_nrow());
+
+    for (int h = 0; h < nirrep_; ++h) {
+        if (!rowspi_[h]) continue;
+
+        // Dimension
+        int n = rowspi_[h];
+        // Pivot
+        pivot[h].resize(n);
+
+        // Perform decomposition. Since the matrix is symmetric, row vs column major ordering makes no difference.
+        int rank;
+        int err = C_DPSTRF('L', rowspi_[h], matrix_[h][0], rowspi_[h], pivot[h].data(), &rank, tol, work.data());
+        nchol[h] = rank;
+
+        // Fix pivots, Fortran to C
+        pivot[h].resize(rank);
+        for (int i = 0; i < rank; i++) pivot[h][i]--;
+
+        if (err != 0) {
+            if (err < 0) {
+                outfile->Printf("pivoted_cholesky: C_DPSTRF: argument %d has invalid parameter.\n", -err);
+
+                abort();
+            }
+            if (err > 1) {
+                outfile->Printf(
+                    "pivoted_cholesky: C_DPSTRF error code %d: the matrix A is either rank deficient with computed "
+                    "rank as returned in RANK, or is not positive semidefinite.",
+                    err);
+                abort();
+            }
+        }
+    }
+
+    // Properly sized return matrix
+    auto L = std::make_shared<Matrix>("Cholesky decomposed matrix", nirrep_, rowspi_, nchol);
+    L->zero();
+    for (int h = 0; h < nirrep_; ++h) {
+        for (int m = 0; m < rowspi_[h]; ++m) {
+            for (int n = 0; n < std::min(m + 1, nchol[h]); ++n) {
+              // Psi4 stores matrices as row major, LAPACK as column major
+              L->set(h, m, n, get(h, n, m));
+            }
+        }
+    }
+    // Switch to the properly sized matrix
```

Comment:
```suggestion
    // Switch to the properly sized matrix
    colspi_ = nchol;
```
This should fix the memory bug - I found it using address sanitizer on linux.  There's another error that's flagged after this fix, but occurs in the DFJK code and is likely a false positive.

Reply:
OK, this is a bug in `matrix.cc`.

Reply:
Fixed in the copy assignment operator.

Reply:
Yes, your fix is better than mine.  I blame a lack of :coffee: